### PR TITLE
Signing and Encryption

### PR DESCRIPTION
**Pull request by sbhatore95**
@kaspth Lets switch to this pull request for future reference and additions. It contains both Signing and Encryption related work :smile: !!

### CODE REVIEW

**Comment by kaspth** [comment]
Hey 👋,

Just a reminder that this won't make it in time for 5.0. This domain is still super complex, so I won't have time to look more into it until after 5.0 is done.

Do look for places to simplify and clarify the code on your own meanwhile. Keep pushing forward! :heart:

File: activesupport/lib/active_support/encryptor/jwe.rb [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+          auth_tag = mac[0...(mac.length / 2)]
+        end
+
+        auth_data << '.' << [jwe_encrypted_key, iv, ciphertext, auth_tag].map { |a| encode a }.join('.')
+      end
+
+      def decrypt(encrypted_message)
+        parts = encrypted_message.split('.')
+        auth_data = parts.shift
+        encrypted_key, iv, ciphertext, auth_tag = parts.map { |a| decode a }
```
Comment (bdewater) [inline activesupport/lib/active_support/encryptor/jwe.rb]: Hi there! Looks like we've been doing similar work (#25874) 😄 you'll want to check the auth_tag length here and add a test for it.

File: activesupport/test/message_encryptor_test.rb [hunk 2]
```diff
@@ -2,8 +2,155 @@
...
+  def test_error_when_message_is_in_invalid_format
+    assert_invalid_format ''
+    assert_invalid_format 'pure.junk.data.in.this.string'
+    assert_invalid_format '.........'
+  end
+
+  def test_messing_with_message_causes_failure
+    h, key, iv, ciphtxt, tag = @encryptor.encrypt_and_sign(@payload, @options).split('.')
+    assert_not_decrypted "#{munge(h)}.#{key}.#{iv}.#{ciphtxt}.#{tag}"
+    assert_not_decrypted "#{h}.#{key}.#{iv}.#{ciphtxt}.#{munge(tag)}"
```
Comment (bdewater) [inline activesupport/test/message_encryptor_test.rb]: Add `assert_not_decrypted "#{h}.#{key}.#{iv}.#{ciphtxt}.#{tag[0..-2]}"`

**Comment by connorshea** [comment]
Any updates on this? This looks awesome :)

**Comment by kaspth** [comment]
Won't be getting time to work on this until after 5.1 😢 — but there's always 5.2! 😄

**Comment by cristianbica** [comment]
@kaspth this would be useful to replace [ActiveStorage::VerifiedKeyWithExpiration](https://github.com/rails/activestorage/blob/master/lib/active_storage/verified_key_with_expiration.rb). Any chances this will make it into 5.2 ?

**Comment by kaspth** [comment]
@cristianbica That's indeed the idea. We're working on it here: https://github.com/rails/rails/pull/29599

@sbhatore thanks for all the work, Siddharth!

**Pull request closed without merging**